File: kotlin.web.demo.server/static/js/File.js
/*
 * Copyright 2000-2014 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Created by Semyon.Atamas on 10/13/2014.
 */
var File = (function () {

    function File(project, content) {
        var instance = {
            toJSON: function () {
                return {
                    name: name,
                    text: text,
                    publicId: publicId
                }
            },
            save: function () {
                if (project.getType() == ProjectType.USER_PROJECT) {
                    save()
                }
            },
            dumpToLocalStorage: function () {
                if (project.getType() != ProjectType.USER_PROJECT) {
                    dumpToLocalStorage();
                }
            },
            loadOriginal: function () {
                fileProvider.loadOriginalFile(instance, setFileData, function () {
                    window.alert("Can't find file origin, maybe it was removed by a user");
                    projectActionsView.setStatus("localFile");
                    revertible = false;
                });
            },
            onFileSaved: function () {
            },
            compareContent: function () {
                if (text == originalText) {
                    modified = false;
                    instance.onUnmodified();
                } else {
                    modified = true;
                    instance.onModified();
                }
            },
            isModified: function () {
                return modified;
            },
            onModified: function () {
            },
            onUnmodified: function () {
            },
            deleteThis: function () {
                project.deleteFile(publicId);
                instance.onDeleted();
            },
            onDeleted: function () {

            },
            makeNotRevertible: function () {
                revertible = false;
            },
            getName: function () {
                return name;
            },
            rename: function (newName) {
                newName = addKotlinExtension(newName);
                name = newName;
                instance.onRenamed(newName);
            },
            onRenamed: function (newName) {
            },
            isModifiable: function () {
                return modifiable;
            },
            isRevertible: function () {
                return revertible;
            },
            getErrors: function () {
                return errors;
            },
            setErrors: function (newErrors) {
                errors = newErrors == null ? [] : newErrors;
            },
            getProjectType: function () {
                return project.getType();
            },
            getText: function () {
                return text;
            },
            getPublicId: function () {
                return publicId;
            },
            setText: function (newText) {
                text = newText;
                instance.compareContent();
            },
            setChangesHistory: function (history) {
                changesHistory = history;
            },
            getChangesHistory: function () {
                return changesHistory;
            },
            getProject: function () {
                return project;
            }
        };

        var name = "";
        var originalText = "";
        var text = "";
        var modifiable = true;
        var errors = [];
        var type = "Kotlin";
        var publicId = "";
        var changesHistory = null;
        var modified = false;
        var revertible = true;

        function save() {
            fileProvider.saveFile(instance, function () {
                originalText = text;
                instance.compareContent();
            });
        }

        function dumpToLocalStorage() {
            localStorage.setItem(publicId, JSON.stringify({
                name: name,
                originalText: originalText,
                text: text,
                publicId: publicId,
                modifiable: modifiable,
                type: type,
                revertible: revertible
            }))
        }

        function setFileData(data) {
            if (data != null) {
                name = data.name;
                text = data.text;
                originalText = data.originalText == null ? text : data.originalText;
                modifiable = data.modifiable;
                publicId = data.publicId;
                type = data.type;
                changesHistory = null;
                revertible = content.hasOwnProperty("revertible") ? content.revertible : true;
            }
        }

        setFileData(content);

        return instance;
    }

    File.EmptyFile = function (project, name, publicId) {
        name = addKotlinExtension(name);
        return new File(project, {
            name: name,
            text: "",
            modifiable: true,
            publicId: publicId
        });
    };

    File.fromLocalStorage = function (project, publicId) {
        return new File(project, JSON.parse(localStorage.getItem(publicId)));
    };

    return File;
})();

File: kotlin.web.demo.frontend/src/File.kt
/*
 * Copyright 2000-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import kotlin.js.dom.html5.localStorage
import kotlin.properties.Delegates

/**
 * Created by Semyon.Atamas on 4/2/2015.
 */

class File(
        val project: Project,
        name: String,
        val id: String,
        text: String = "",
        var originalText: String = text,
        val isModifiable: Boolean = true,
        val type: FileType = FileType.KOTLIN_FILE,
        var isRevertible: Boolean = true
) {
    var errors = array<Error>()
    var changesHistory: dynamic = null;

    val listenableName = VarListener<String>();
    var name: String by Listenable(name, listenableName)

    val listenableIsModified = VarListener<Boolean>()
    var isModified: Boolean by Listenable(false, listenableIsModified)

    var text: String = text
        set(newText: String) {
            isModified = newText != originalText
            $text = newText
        }

    fun toJSON(): dynamic {
        val result = js("({})")
        result.name = name
        result.text = text
        result.publicId = id
        return result
    }

    companion object {

        fun fromJSON(project: dynamic, obj: dynamic) =
                File(project, obj.name, obj.publicId, obj.text, obj.originalText, obj.modifiable, obj.type, obj.revertible)
    }

    //TODO following functions from file to some other class
    fun save() {
        //TODO replace === with == (when == will work correctly)
        if (project.getType() === ProjectType.USER_PROJECT && isModifiable) {
            fileProvider.saveFile(this, {
                originalText = text
                isModified = text != originalText
            });
        }
    }

    fun dumpToLocalStorage() {
        //TODO replace !== with != (when != will work correctly)
        if (project.getType() !== ProjectType.USER_PROJECT) {
            //TODO don't save editor info
            val result = js("({})")
            result.name = name
            result.originalText = originalText
            result.text = text
            result.publicId = id
            result.modifiable = isModifiable
            result.type = type
            result.revertible = isRevertible
            localStorage.set(id, JSON.stringify(result));
        }
    }

    fun loadOriginal(){

    }
}
